### PR TITLE
User routes

### PR DESCRIPTION
**Pull request by kskkido**
created backend user routes
- needs testing

### CODE REVIEW

File: server/users.js [hunk 1]
```diff
@@ -2,30 +2,117 @@
...
+})
+
+.post((req, res, next) => {
+  User.create(req.body)
+  .then(user => res.status(201).json(user))
+  .catch(next)
+})
+
+router.route('/:userId')
+
```
Comment (emtseng) [inline server/users.js]: style: would remove space here

File: server/users.js [hunk 2]
```diff
@@ -2,30 +2,117 @@
...
+.get(mustBeLoggedIn, (req, res, next) => {
+  res.json(req.targetUser)
+})
+
```
Comment (emtseng) [inline server/users.js]: would also remove space here

Comment (emtseng) [reply server/users.js]: would actually remove spaces throughout

File: server/users.js [hunk 3]
```diff
@@ -2,30 +2,117 @@
...
+  .then(() => res.sendStatus(204))
+  .catch(next)
+})
+
+/*  -------------- ADDRESSES ------------- */
+
+router.route('/:userId/:addressType')
+
+.get(mustBeLoggedIn, (req, res, next) => {
+  req.targetUser['get' + req.addressMethod]()
```
Comment (emtseng) [inline server/users.js]: watch your variable names -- believe this needs to be req.addressType as defined in your router.param up top

**Review by emtseng — CHANGES_REQUESTED** [review]
This is really neat! One important variable name change needed to make the address method/type available after your router.param, but otherwise this looks good.

File: server/users.js [hunk 4]
```diff
@@ -1,31 +1,123 @@
...
-      .catch(next))
-  .get('/:id',
-    mustBeLoggedIn,
-    (req, res, next) =>
-      User.findById(req.params.id)
-      .then(user => res.json(user))
-      .catch(next))
+module.exports = router
+
+router.param(':userId', (req, res, next, userId) => {
```
Comment (emtseng) [inline server/users.js]: believe the first argument of the .param should just be 'userId' with no :

File: server/users.js [hunk 5]
```diff
@@ -1,31 +1,123 @@
...
+  User.find({
+    where: {id: userId},
+    include: [
+      {model: Address, as: 'shippingAddresses'},
```
Comment (emtseng) [inline server/users.js]: you include Address in your default scope, I believe -- no reason to eager load them here

File: server/users.js [hunk 6]
```diff
@@ -1,31 +1,123 @@
...
+    ]
+  })
+  .then(targetUser => {
+    if (!targetUser) return res.sendStatus(404)
+    req.targetUser = targetUser
+  })
+  .catch(next)
+})
+
+router.param(':addressType', (req, res, next, addressType) => {
```
Comment (emtseng) [inline server/users.js]: see above comment -- router.param('addressType' ... ), no colon

File: server/users.js [hunk 7]
```diff
@@ -1,31 +1,123 @@
...
+router.route('/')
+  // The forbidden middleware will fail *all* requests to list users.
+  // Remove it if you want to allow anyone to list all users on the site.
+  //
+  // If you want to only let admins list all the users, then you'll
+  // have to add a role column to the users table to support
+  // the concept of admin users.
+  .get(forbidden('listing users is not allowed'), (req, res, next) => {
+    User.findAll({
+      include: [
```
Comment (emtseng) [inline server/users.js]: see above on scope and eager loading

File: server/users.js [hunk 8]
```diff
@@ -1,31 +1,123 @@
...
+  .delete(mustBeLoggedIn, (req, res, next) => {
+    req.targetUser['remove' + req.addressType]()
+    .then(() => res.sendStatus(204))
+    .catch(next)
+  })
+
+/* --------- INDIVIDUAL ADDRESS ----------- */
+
+router.route('/:userId/:addressType/:addressId')
+  .get(mustBeLoggedIn, (req, res, next) => {
```
Comment (emtseng) [inline server/users.js]: style -- in code review believe we discussed not indenting here

**Review by emtseng — CHANGES_REQUESTED** [review]
just a couple of issues with router.param and one comment on style

File: db/seed.js [hunk 9]
```diff
@@ -78,13 +78,13 @@ const addresses = seed(Address, createAddressList())
...
+  Shirt: ['HIGH NECK HALF SLEEVE T-SHIRT', 'DRY YOGA SHORT-SLEEVE T-SHIRT', 'EXTRA FINE MERINO V-NECK CARDIGAN', 'PEANUTS SHORT-SLEEVE GRAPHIC TEE', 'EASY CARE OXFORD LONG SLEEVE SHIRT'],
+  Pants: ['RELAXED ANKLE PANTS', 'SKINNY STRAIGHT JEANS', 'AIRISM PILE LINED LOUNGE SHORTS'],
+  Shoes: ['DISNEY MAROON SLIPPERS', 'NIKE AIR MAX 90 ULTRA 2.0 BREATHE']
 }
 const descriptions = ['is fabulous!', 'is great for any occasion.', 'is binary', 'is classy and elegant.', 'features a soft, fluffy feel!', 'has a gently snug, relaxed fit', 'is ideal for relaxing at home.']
 const gender = ['MEN', 'WOMEN']
-const clothingType = ['shirt', 'pants', 'shoes']
+const clothingType = ['Shirt', 'Pants', 'Shoes']
 const photos = {
   shirt: 'http://demandware.edgesuite.net/aawj_prd/on/demandware.static/-/Library-Sites-CTShirtsSharedLibrary/default/dw9877fa95/images/plp-tooltip_fit-shirt-extraslim.jpg',
```
Comment (tinaonthelam) [inline db/seed.js]: capitalize keys

File: db/seed.js [hunk 10]
```diff
@@ -97,12 +97,12 @@ function createProduct() {
...
-    title: sex + ' ' + product,
+    name: sex + ' ' + product,
     description: product + ' ' + description,
     price: Math.floor(Math.random() * 50 + 10),
-    gender: sex,
-    clothingType: clothType,
-    photo: photos[clothingType]
+    inventory: 50,
+    categories: [sex, clothType, 'Featured'],
+    photos: photos[clothType]
```
Comment (tinaonthelam) [inline db/seed.js]: photos takes an array

File: db/models/user.js [hunk 11]
```diff
@@ -27,10 +27,12 @@ module.exports = db => db.define('users', {
   indexes: [{fields: ['email'], unique: true}],
 
   scopes: {
```
Comment (tinaonthelam) [inline db/models/user.js]: default scope *

File: server/addresses.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+router.param('addressType', (req, res, next, addressType) => {
+  if (addressType !== 'billingAddress' || addressType !== 'shippingAddress') {
+    res.sendStatus(404).end()
+  }
+  req.addressType = addressType
+  next()
+})
+
+router.route(':addressType')
```
Comment (tinaonthelam) [inline server/addresses.js]: /:addressType

File: server/users.js [hunk 13]
```diff
@@ -1,31 +1,66 @@
...
+router.param('userId', (req, res, next, userId) => {
+  User.find({where: {id: userId}})
+  .then(targetUser => {
+    if (!targetUser) return res.sendStatus(404)
+    req.targetUser = targetUser
+  })
+  .catch(next)
+})
+
+router.param('addressType', (req, res, next, addressType) => {
```
Comment (tinaonthelam) [inline server/users.js]: remove this

File: server/addresses.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+})
+.delete(mustBeLoggedIn, (req, res, next) => {
+  req.targetUser['remove' + req.addressType]()
+  .then(() => res.sendStatus(204))
+  .catch(next)
+})
+
+/* --------- INDIVIDUAL ADDRESS ----------- */
+
+router.route(':addressType/:addressId')
```
Comment (tinaonthelam) [inline server/addresses.js]: /:addressType/:addressId

**Pull request merged**